configuration/purging-serverside.md

```diff
@@ -0,0 +1,208 @@
+
+## How purging works client-side
+
+Purging runs on the user's device at startup, before the application booting, when any of these 
+conditions are met:
+
+- The device was just setup and so initial sync just occurred.
+- The user's roles list has changed
+- It has been more than 7 days since the user has purged.
+
+Purging will not run on startup if purge is not configured, if last purge occurred recently or if 
+the user doesn't have an internet connection. 
+
+When purging runs, the device calls an API endpoint that returns batches of doc ids that have
+been purged since last time the same device has run purge.
+The system is similar to CouchDB replication, in the sense that a `checkpointer` document 
+is saved in the corresponding server-side database, that stores the `last_seq` that the device has 
+downloaded and is used to get the next batch of ids.
+After receiving a batch of ids, the device simply deletes the indicated docs locally, 
+marking them with a `purged` flag. 
+
+## Purging function configuration
+
+To enable purging, write your purging function to `purging.js` in your project root:
+
+```js
+function(userCtx, contact, reports, messages) {
+  const old = Date.now() - (1000 * 60 * 60 * 24 * 365);
+  const oldMessages = Date.now() - (1000 * 60 * 60 * 24 * 90);
+
+  const reportsToPurge = reports
+                           .filter(r => r.reported_date < old)
+                           .map(r => r._id);
+  const messagesToPurge = messages
+                           .filter(m => m.reported_date < oldMessages)
+                           .map(m => m._id);
+
+  return [...reportsToPurge, ...messagesToPurge];
+}
+```
+
+As shown above, you should be declaring an _anonymous_ function (with no name).
+
+This function takes four parameters:
+ - `userCtx`, an object with the user's `roles` as fields. For more information read the [documentation for the User Context Object](https://docs.couchdb.org/en/stable/json-structure.html#userctx-object).
+ - `contact`, the contact document of a patient or other contact who has reports about them.
+ - `reports`, an array of all reports for that patient that are present on the server.
+ - `messages`, an array of sms messages that the contact has sent or received
+
+And should return an array of `_id` values for docs you would like to be purged 
+(or `undefined` / nothing if you don't wish to purge anything). 
+Only ids of docs that were passed to the function are valid for purging: you are not allowed to purge other documents.
+
+In the cases of reports that do not have patients or their patients are not found, the `purge` function
+will receive an empty object as `contact`. 
+In the cases of reports about deleted patients, the `purge` function will receive a `{ _deleted: true }`
+object as the `contact`. 
+
+## Other configuration
+
+You must set a schedule for purging to run server-side.
```

Comment:
I want to preface this with I'm not sure this would ever happen outside of QA/dev. It seems that this isn't running on an actual cron style schedule. It runs after the 5 minute sentinel check plus whatever the schedule is. So if someone had it set to every minute  it wouldn't actually run every minute. It would run every 6 minutes.

Reply:
Initially, I had it running on an actual schedule and later changed it so it would run at most once per Sentinel scheduler "tick" (it saved me from having to stop and start the process in e2e tests). 
I can roll it back and allow it to run on an interval instead of a timeout (so it could execute multiple times per "tick"). I have to revisit how purging is "triggered" in e2e tests anyway, my initial plan (of using Stefan's Sentinel restart script) has flunk as the code doesn't work properly on Travis.